### PR TITLE
Warp Affine tensor HIP

### PR DESCRIPTION
**Pull request by r-abishek**
- Adds support for warp affine tensor on hip.
- Support for U8/F32/F16/I8, NCHW/NHWC/toggle, Bilinear/NearestNeighbor is added.
- Adds corresponding unit tests and performance tests.

### CODE REVIEW

**Comment by kiritigowda** [comment]
@r-abishek failing the code analysis - [details](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/rpp/pullRequest?prid=9009296&bid=28008902)

**Comment by r-abishek** [comment]
@kiritigowda Yes, actually I checked it out it. It says - Variable 'src_f6.x' is reassigned before using the previous assignment, but I do use using the previous assignment from L1763-L1768 (https://github.com/GPUOpen-ProfessionalCompute-Libraries/rpp/blob/baa063da95b36fb307ccaf71b30ac5f730973048/src/include/hip/rpp_hip_common.hpp#L1763-L1768)
Not sure if codacy understood the pointer cast on L1757

**Comment by kiritigowda** [comment]
> @kiritigowda Yes, actually I checked it out it. It says - Variable 'src_f6.x' is reassigned before using the previous assignment, but I do use using the previous assignment from L1763-L1768 (
> 
> https://github.com/GPUOpen-ProfessionalCompute-Libraries/rpp/blob/baa063da95b36fb307ccaf71b30ac5f730973048/src/include/hip/rpp_hip_common.hpp#L1763-L1768
> 
> )
> Not sure if codacy understood the pointer cast on L1757

Thanks, I will close the issue.

File: src/include/hip/rpp_hip_common.hpp [hunk 1]
```diff
@@ -1481,4 +1618,420 @@ __device__ __forceinline__ void rpp_hip_math_subtract24_const(d_float24 *src_f24
...
+    srcNeighborhood_f12->x.x = src_f6f3s->x.x;
+    srcNeighborhood_f12->x.y = src_f6f3s->y.x;
+    srcNeighborhood_f12->y.x = src_f6f3s->x.y;
+    srcNeighborhood_f12->y.y = src_f6f3s->y.y;
+    srcNeighborhood_f12->z.x = src_f6f3s->x.z;
+    srcNeighborhood_f12->z.y = src_f6f3s->y.z;
+    srcIdx += srcStrideH;
+    src_h6 = *(d_half6 *)&srcPtr[srcIdx];
+    src_f6.x = __half22float2(src_h6.x);
+    src_f6.y = __half22float2(src_h6.y);
```
Comment (rrawther) [inline src/include/hip/rpp_hip_common.hpp:1772]: change to what we have discussed

Comment (r-abishek) [reply src/include/hip/rpp_hip_common.hpp:2026]: Added the fix for this issue. Will be checking on the xyz notation changes in a separate pr.

**Review by rrawther — APPROVED** [review]
@paveltc to verify before merge

File: src/modules/hip/kernel/warp_affine.hpp [hunk 2]
```diff
@@ -0,0 +1,481 @@
+#include <hip/hip_runtime.h>
+#include "hip/rpp_hip_common.hpp"
+
+// -------------------- Set 0 - warp_affine device helpers --------------------
+
+__device__ void warp_affine_srclocs_hip_compute(float affineMatrixElement, float4 locSrcComponent_f4, d_float8 *locSrcPtr_f8)
+{
+    d_float8 increment_f8;
+    increment_f8.x.x = 0;
```
Comment (asalmanp) [inline src/modules/hip/kernel/warp_affine.hpp:9]: use make_float4 and fill the increment_f8.x all at once.

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: src/include/hip/rpp_hip_common.hpp [hunk 3]
```diff
@@ -822,30 +850,30 @@ __device__ __forceinline__ void rpp_hip_load24_pkd3_and_unpack_to_float24_pln3(f
...
-    src_f24->y.x.w = srcPtr_f24->y.x.z;
-    src_f24->z.x.w = srcPtr_f24->y.x.w;
     src_f24->x.y.x = srcPtr_f24->y.y.x;
-    src_f24->y.y.x = srcPtr_f24->y.y.y;
-    src_f24->z.y.x = srcPtr_f24->y.y.z;
     src_f24->x.y.y = srcPtr_f24->y.y.w;
+    src_f24->x.y.z = srcPtr_f24->z.x.z;
+    src_f24->x.y.w = srcPtr_f24->z.y.y;
 
+    src_f24->y.x.x = srcPtr_f24->x.x.y;
```
Comment (asalmanp) [inline src/include/hip/rpp_hip_common.hpp:861]: change line 861-864 to a single line src_f24->y.x = make_float4(srcPtr_f24->x.x.y, srcPtr_f24->x.y.x, srcPtr_f24->x.y.w, srcPtr_f24->y.x.z);

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: src/include/hip/rpp_hip_common.hpp [hunk 4]
```diff
@@ -822,30 +850,30 @@ __device__ __forceinline__ void rpp_hip_load24_pkd3_and_unpack_to_float24_pln3(f
...
+    src_f24->y.y.x = srcPtr_f24->y.y.y;
     src_f24->y.y.y = srcPtr_f24->z.x.x;
-    src_f24->z.y.y = srcPtr_f24->z.x.y;
-    src_f24->x.y.z = srcPtr_f24->z.x.z;
     src_f24->y.y.z = srcPtr_f24->z.x.w;
-    src_f24->z.y.z = srcPtr_f24->z.y.x;
-    src_f24->x.y.w = srcPtr_f24->z.y.y;
     src_f24->y.y.w = srcPtr_f24->z.y.z;
+
+    src_f24->z.x.x = srcPtr_f24->x.x.z;
```
Comment (asalmanp) [inline src/include/hip/rpp_hip_common.hpp:870]: similar to the above comment, assign src_f24->z.x as a float4. please make a similar change for src_f24->z.y

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: src/include/hip/rpp_hip_common.hpp [hunk 5]
```diff
@@ -1330,40 +1354,41 @@ __device__ __forceinline__ void rpp_hip_pack_float24_pln3_and_store24_pkd3(uchar
 
 __device__ __forceinline__ void rpp_hip_pack_float24_pln3_and_store24_pkd3(float *dstPtr, d_float24 *dst_f24)
 {
-    d_float24 *dstPtr_f24;
-    dstPtr_f24 = (d_float24 *)dstPtr;
+    d_float24 dstPtr_f24;
+
+    dstPtr_f24.x.x.x = dst_f24->x.x.x;
```
Comment (asalmanp) [inline src/include/hip/rpp_hip_common.hpp:1359]: also change this function to assign dstPtr_f24.x.x, dstPtr_f24.x.y, ... as float4

**Review by asalmanp — CHANGES_REQUESTED** [review]

**Comment by paveltc** [comment]
@asalmanp @rrawther @r-abishek 
I ran the tests for this PR and everything looks good.
The only possible concern is that some of the images have grey backgrounds instead of black. Other than that, everything looks as it should.

File: src/modules/hip/kernel/warp_affine.hpp (hunk 2 above)
Comment (r-abishek) [reply src/modules/hip/kernel/warp_affine.hpp]: Done

File: src/include/hip/rpp_hip_common.hpp (hunk 3 above)
Comment (r-abishek) [reply src/include/hip/rpp_hip_common.hpp]: Done

File: src/include/hip/rpp_hip_common.hpp (hunk 4 above)
Comment (r-abishek) [reply src/include/hip/rpp_hip_common.hpp]: Done

File: src/include/hip/rpp_hip_common.hpp (hunk 5 above)
Comment (r-abishek) [reply src/include/hip/rpp_hip_common.hpp]: Done

**Comment by r-abishek** [comment]
@paveltc Thanks! The grey backgrounds are just for i8 due to the range from -128 to 127, which is usually the case for all rpp functions.

**Comment by r-abishek** [comment]
@asalmanp @rrawther I've made the changes for the make_float4 in all applicable cases. I've also added some comments to improve readability - basically in load/store helper functions to indicate which pixel-writes are happening and in what layout (pln3/pkd3) - https://github.com/GPUOpen-ProfessionalCompute-Libraries/rpp/pull/96/commits/789abc016f2a53d7a5f500dd01ad779fdf622023

**Pull request merged**